heapster-saw/src/Verifier/SAW/Heapster/Implication.hs

```diff
@@ -6698,194 +6666,268 @@ solveForArrayH ps _ mb_ap =
+-- the required offset and length or that includes them in its range, or is a
+-- block permission that that includes the required offset and length in its
+-- range. Currently, there are no rules for changing the stride of an array, so
+-- arrays with different strides are not considered. If no such permission is
+-- found on the left, then a fully borrowed array permission is created, where
+-- the borrows are calculated to either line up with the ranges of permissions
+-- we already hold on the left, so they can be returned, or to be in the desired
+-- output permission, so we do not have to return them.
+--
+-- In all of these cases, an array permission with the required offset and
+-- length is either found on the left or created, and all cases then reduce to
+-- case 1. At this point, the algorithm equalizes the borrows, meaning that it
+-- returns any borrows on the left that are not on the right (where the right is
+-- the desired output permission) and borrows any borrows on the right that are
+-- not on the left. It then adjusts the read/write and lifetime modalities and
+-- coerces the cell permissions if necessary. These steps are performed by the
+-- helper function 'proveVarLLVMArray_FromArray'.
 proveVarLLVMArrayH ::
   (1 <= w, KnownNat w, NuMatchingAny1 r) => ExprVar (LLVMPointerType w) ->
-  Bool -> PartialSubst vars -> [AtomicPerm (LLVMPointerType w)] ->
+  PartialSubst vars -> [AtomicPerm (LLVMPointerType w)] ->
   Mb vars (LLVMArrayPerm w) ->
   ImplM vars s r (ps :> LLVMPointerType w) (ps :> LLVMPointerType w) ()
 
 -- Special case: if the length is 0, prove an empty array
-proveVarLLVMArrayH x _ psubst ps mb_ap
+proveVarLLVMArrayH x psubst ps mb_ap
   | Just len <- partialSubst psubst $ mbLLVMArrayLen mb_ap
   , bvIsZero len =
     recombinePerm x (ValPerm_Conj ps) >>>
     partialSubstForceM mb_ap "proveVarLLVMArray: incomplete psubst" >>>= \ap ->
     implLLVMArrayEmpty x ap
 
--- Special case: if we already have a single array permission that covers the
--- RHS, then just use (a subset of) that
-proveVarLLVMArrayH x _p psubst ps mb_ap
+-- If we have a single array permission that covers the RHS, then we are in case
+-- 1 or 2, so either use that or borrow or copy a portion of it and proceed to
+-- proveVarLLVMArray_FromArray
+proveVarLLVMArrayH x psubst ps mb_ap
   | Just off <- partialSubst psubst $ mbLLVMArrayOffset mb_ap
   , Just len <- partialSubst psubst $ mbLLVMArrayLen mb_ap
   , Just lenBytes <- partialSubst psubst $ mbLLVMArrayLenBytes mb_ap
-  , Just i   <- findIndex (containsRHS off lenBytes) ps
+  , stride <- mbLLVMArrayStride mb_ap
+  , Just i   <- findIndex (suitableAP off lenBytes stride) ps
   , Perm_LLVMArray ap_lhs <- ps!!i =
-      implGetConjM x ps i >>>= \ps' ->
-      recombinePerm x (ValPerm_Conj ps') >>>
+    implVerbTraceM (\info -> pretty "proveVarLLVMArrayH case 1: using" <+>
+                             permPretty info ap_lhs) >>>
```

Comment:
Mention this is case 1 or 2, or perhaps modify the case list in the initial comment to match the cases of the pattern match?